src/Symfony/Component/DependencyInjection/Tests/Fixtures/Prototype/NotFoo.php

```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace Symfony\Component\DependencyInjection\Tests\Fixtures\Prototype;
+
+use Symfony\Component\DependencyInjection\Attribute\NotWhen;
+
+#[NeverInProduction]
+#[NotWhen(env: 'dev')]
```

Comment:
To properly test this case, we would need several tests involving it:

- one test using the production env and asserting that the service is not there
- one test using the dev env and asserting that the service is not there (we need both tests to ensure that the repeated attributes are properly taken into account)
- one test using another env and asserting that the service is there.

Otherwise, the tests are not actually ensuring that the implementation works fine (and this is shown by the fact that the current implementation passes your tests but does not work fine)

Reply:
I missed the test covering `#[When]`, my bad. Cases added now 👍